### PR TITLE
Zoom and pan to selection from multiple layers

### PR DESCRIPTION
**Pull request by TurboGraphxBeige**
### Zoom and pan to the selected features from all the selected layers in the Layers panel.

Implements the feature request https://github.com/qgis/QGIS/issues/38668, and more.

**Zoom**
----
- A menu item "Zoom to Selection from All Layers" has been added inside the "View" menu. This action zooms the map to the combined extent of the selected features from all the layers in the map.
- The "Zoom to Selection" button in the Map Navigation Toolbar zooms to the selected features of all the selected layers in the layer tree. 
- The "Zoom to Selection" context menu action in the layer tree zooms to the selected features of all the selected layers in the layer tree.

To achieve that:

- A new function QgsMapCanvas::zoomToAllSelected has been created that takes a list of QgsMapLayer as input. 
- QgsApp::zoomToSelected and QgsLayerTreeViewDefaultActions::zoomToSelection have been modified to execute the existing QgsMapCanvas::zoomToSelected if a single layer is selected in the layer tree or QgsMapCanvas::zoomToAllSelected if more than one layer is selected in the layer tree. 
- QgsApp::zoomToAllSelected() has been created for the "View" menu item "Zoom to Selection from All Layers". 
- QgsAppLayerTreeViewMenuProvider::createContextMenu has been modified to enable the menu item "Zoom to Selection" when at least one selected layer in the layer tree has a selected feature.

Zoom-to-selection4

**Pan**
----
- A menu item "Pan Map to Selection from All Layers" has been added inside the "View" menu. This action pans the map to the combined extent of the selected features from all the layers in the map.
- The "Pan to Selection" button in the Map Navigation Toolbar pans the map to the selected features of all the selected layers in the layer tree. 

To achieve that:
- A new function "QgsMapCanvas::panToAllSelected" has been created that takes a list of QgsMapLayer as input.
- QgsApp::panToSelected() has been modified to execute the existing QgsMapCanvas::panToSelected if a single layer is selected in the layer tree or QgsMapCanvas::panToAllSelected if more than one layer is selected in the layer tree. 
- QgsApp::panToAllSelected() has been created for the "View" menu item "Pan Map to Selection from All Layers".

Pan-map2

### CODE REVIEW

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 1]
```diff
@@ -155,8 +155,28 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
         menu->addAction( zoomToLayer );
         if ( vlayer )
         {
+          QList<QgsMapLayer *> selectedLayers = mView->selectedLayers();
+          bool hasSelectedFeature;
+          for ( int i = 0; i < selectedLayers.size(); ++i )
+          {
+            QgsMapLayer *layer = selectedLayers.at( i );
+            QgsVectorLayer *vLayer;
```
Comment (m-kuhn) [inline src/app/qgsapplayertreeviewmenuprovider.cpp:163]:
```suggestion
```

**Comment by roya0045** [comment]
Why the changes to the prepare commit file?

**Comment by TurboGraphxBeige** [comment]
> Why the changes to the prepare commit file?

I messed up. That's not intentional. I was editing .git/hooks/pre-commit but I did it wrong. It's too early in the morning for me I guess.

**Comment by m-kuhn** [comment]
Very sorry for the steep learning curve. You'll get the hang of it soon

**Comment by roya0045** [comment]
@TurboGraphxBeige No issues, wanted to make sure it didn't slip through.

**Comment by TurboGraphxBeige** [comment]
Morning goof fixed. All checks passed. Yay!

**Comment by nyalldawson** [comment]
@TurboGraphxBeige I like where you're going with this! My one reservation is about the new menu items being added to some already very crowded menus 😝

Do we really need these? I think we could safely default to using the new behaviour with the existing actions without any user harm.

@nirvn what do you think?

**Comment by nirvn** [comment]
+1 to the behavior addition, -1 to the action addition :) IMHO, we can apply the logic onto the current action and avoid overcrowding menu (and overwhelming the users) with multiple actions here. If a user has selected multiple layers in the context of feature selection and clicks on that action, 99.9% of the time he/she would wishes for this :)

**Comment by TurboGraphxBeige** [comment]
Totally agree with you @nyalldawson and @nirvn. Looking back at it again after a few days makes those menu actions look a bit out of place. They are useful but maybe not at the expanse of crowding even more those menus.

File: src/gui/qgsmapcanvas.h [hunk 2]
```diff
@@ -310,6 +311,12 @@ class GUI_EXPORT QgsMapCanvas : public QGraphicsView
     //! Pan to the selected features of current (vector) layer keeping same extent.
     void panToSelected( QgsVectorLayer *layer = nullptr );
 
+    /**
+     * Pan to the combined extent of the selected features of all provided (vector) layers.
+     * \param layers A list of layers
+     */
+    void panToAllSelected( const QList<QgsMapLayer *> *layers );
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.h:318]:
```suggestion
    void panToSelected( const QList<QgsMapLayer *> &layers );
```

File: src/gui/qgsmapcanvas.h [hunk 3]
```diff
@@ -310,6 +311,12 @@ class GUI_EXPORT QgsMapCanvas : public QGraphicsView
     //! Pan to the selected features of current (vector) layer keeping same extent.
     void panToSelected( QgsVectorLayer *layer = nullptr );
 
+    /**
+     * Pan to the combined extent of the selected features of all provided (vector) layers.
+     * \param layers A list of layers
+     */
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.h:317]:
```suggestion
     * \since QGIS 3.18
     */
```

File: src/gui/qgsmapcanvas.h [hunk 4]
```diff
@@ -865,6 +872,12 @@ class GUI_EXPORT QgsMapCanvas : public QGraphicsView
      */
     void zoomToSelected( QgsVectorLayer *layer = nullptr );
 
+    /**
+     * Zoom to the combined extent of the selected features of all provided (vector) layers.
+     * \param layers A list of layers
+     */
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.h:878]:
```suggestion
     * \since QGIS 3.18
     */
```

File: src/gui/qgsmapcanvas.h [hunk 5]
```diff
@@ -865,6 +872,12 @@ class GUI_EXPORT QgsMapCanvas : public QGraphicsView
...
+    void zoomToAllSelected( const QList<QgsMapLayer *> *layers );
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.h:879]:
```suggestion
    void zoomToSelected( const QList<QgsMapLayer *> &layers );
```

File: src/app/qgisapp.cpp [hunk 6]
```diff
@@ -8191,12 +8191,25 @@ void QgisApp::zoomOut()
 
 void QgisApp::zoomToSelected()
 {
-  mMapCanvas->zoomToSelected();
+  QList<QgsMapLayer *> layers = mLayerTreeView->selectedLayers();
+
+  if ( layers.size() > 1 && !layers.isEmpty() )
+    mMapCanvas->zoomToAllSelected( &layers );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:8197]:
```suggestion
    mMapCanvas->zoomToSelected( layers );
```

File: src/app/qgisapp.cpp [hunk 7]
```diff
@@ -8191,12 +8191,25 @@ void QgisApp::zoomOut()
 
 void QgisApp::zoomToSelected()
 {
-  mMapCanvas->zoomToSelected();
+  QList<QgsMapLayer *> layers = mLayerTreeView->selectedLayers();
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:8194]:
```suggestion
  const QList<QgsMapLayer *> layers = mLayerTreeView->selectedLayers();
```

File: src/app/qgisapp.cpp [hunk 8]
```diff
@@ -8191,12 +8191,25 @@ void QgisApp::zoomOut()
...
+
+  if ( layers.size() > 1 && !layers.isEmpty() )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:8196]:
```suggestion
  if ( layers.size() > 1 )
```

(can't be empty if it has > 1 items :wink: )

File: src/app/qgisapp.cpp [hunk 9]
```diff
@@ -8191,12 +8191,25 @@ void QgisApp::zoomOut()
...
+
+  else
+    mMapCanvas->zoomToSelected();
+
 }
 
 void QgisApp::panToSelected()
 {
-  mMapCanvas->panToSelected();
+  QList<QgsMapLayer *> layers = mLayerTreeView->selectedLayers();
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:8206]:
```suggestion
  const QList<QgsMapLayer *> layers = mLayerTreeView->selectedLayers();
```

File: src/app/qgisapp.cpp [hunk 10]
```diff
@@ -8191,12 +8191,25 @@ void QgisApp::zoomOut()
...
+
+  if ( layers.size() > 1 && !layers.isEmpty() )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:8208]:
```suggestion
  if ( layers.size() > 1 )
```

File: src/app/qgisapp.cpp [hunk 11]
```diff
@@ -8191,12 +8191,25 @@ void QgisApp::zoomOut()
...
+    mMapCanvas->panToAllSelected( &layers );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:8209]:
```suggestion
    mMapCanvas->panToSelected( layers );
```

File: src/app/qgisapp.cpp [hunk 12]
```diff
@@ -8191,12 +8191,25 @@ void QgisApp::zoomOut()
...
+
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:8210]:
```suggestion
```

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 13]
```diff
@@ -156,8 +156,28 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
         menu->addAction( zoomToLayers );
         if ( vlayer )
         {
+          QList<QgsMapLayer *> selectedLayers = mView->selectedLayers();
+          bool hasSelectedFeature;
+          for ( int i = 0; i < selectedLayers.size(); ++i )
+          {
+            QgsMapLayer *layer = selectedLayers.at( i );
```
Comment (nyalldawson) [inline src/app/qgsapplayertreeviewmenuprovider.cpp:163]:
```suggestion
          const QList<QgsMapLayer *> selectedLayers = mView->selectedLayers();
          bool hasSelectedFeature = false;
          for ( const QgsMapLayer *layer : selectedLayers )
          {
```

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 14]
```diff
@@ -156,8 +156,28 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
...
+
+            if ( layer->type() == QgsMapLayerType( 0 ) )
+            {
+              QgsVectorLayer *vLayer = qobject_cast<QgsVectorLayer *>( layer );
```
Comment (nyalldawson) [inline src/app/qgsapplayertreeviewmenuprovider.cpp:167]:
```suggestion
            if ( const QgsVectorLayer *vLayer = qobject_cast<const QgsVectorLayer *>( layer ) )
            {
```

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 15]
```diff
@@ -156,8 +156,28 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
...
+
+              if ( !vLayer->selectedFeatureIds().isEmpty() )
```
Comment (nyalldawson) [inline src/app/qgsapplayertreeviewmenuprovider.cpp:169]:
```suggestion
              if ( vLayer->selectedFeatureCount() > 0 )
```

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 16]
```diff
@@ -156,8 +156,28 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
...
+              {
+                hasSelectedFeature = true;
+                break;
+              }
+              else
+                hasSelectedFeature = false;
```
Comment (nyalldawson) [inline src/app/qgsapplayertreeviewmenuprovider.cpp:175]:
```suggestion
```

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp [hunk 17]
```diff
@@ -312,11 +312,19 @@ void QgsLayerTreeViewDefaultActions::zoomToLayers( QgsMapCanvas *canvas )
 
 void QgsLayerTreeViewDefaultActions::zoomToSelection( QgsMapCanvas *canvas )
 {
-  QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( mView->currentLayer() );
-  if ( !layer )
-    return;
+  QgsVectorLayer *layer;
+
+  if ( mView->currentLayer()->type() == QgsMapLayerType( 0 ) )
+    layer = qobject_cast<QgsVectorLayer *>( mView->currentLayer() );
```
Comment (nyalldawson) [inline src/gui/layertree/qgslayertreeviewdefaultactions.cpp:318]:
```suggestion
  QgsVectorLayer *layer = qobject_cast< QgsVectorLayer * >( mView->currentLayer() );
```

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp [hunk 18]
```diff
@@ -312,11 +312,19 @@ void QgsLayerTreeViewDefaultActions::zoomToLayers( QgsMapCanvas *canvas )
...
+
+  QList<QgsMapLayer *> layers = mView->selectedLayers();
+
+  if ( layers.size() > 1 && !layers.isEmpty() )
+    canvas->zoomToAllSelected( &layers );
+
+  else if ( layers.size() <= 1 && layer )
+    canvas->zoomToSelected( layer );
```
Comment (nyalldawson) [inline src/gui/layertree/qgslayertreeviewdefaultactions.cpp:326]:
```suggestion
  const QList<QgsMapLayer *> layers = mView->selectedLayers();

  if ( layers.size() > 1 )
    canvas->zoomToSelected( layers );
  else if ( layers.size() <= 1 && layer )
    canvas->zoomToSelected( layer );
```

File: src/gui/qgsmapcanvas.cpp [hunk 19]
```diff
@@ -1343,12 +1344,50 @@ void QgsMapCanvas::zoomToSelected( QgsVectorLayer *layer )
...
   // also check that rect is empty, as it might not in case of multi points
   if ( layer->geometryType() == QgsWkbTypes::PointGeometry && rect.isEmpty() )
-  {
     rect = optimalExtentForPointLayer( layer, rect.center() );
-  }
+
   zoomToFeatureExtent( rect );
 }
 
+void QgsMapCanvas::zoomToAllSelected( const QList<QgsMapLayer *> *layers )
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.cpp:1352]:
```suggestion
void QgsMapCanvas::zoomToSelected( const QList<QgsMapLayer *> &layers )
```

File: src/gui/qgsmapcanvas.cpp [hunk 20]
```diff
@@ -1343,12 +1344,50 @@ void QgsMapCanvas::zoomToSelected( QgsVectorLayer *layer )
...
+{
+  QgsVectorLayer *layer;
+
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.cpp:1355]:
```suggestion
```

File: src/gui/qgsmapcanvas.cpp [hunk 21]
```diff
@@ -1343,12 +1344,50 @@ void QgsMapCanvas::zoomToSelected( QgsVectorLayer *layer )
...
+  QgsRectangle rect;
+  rect.setMinimal();
+  QgsRectangle selectionExtent;
+  selectionExtent.setMinimal();
+
+  for ( int i = 0; i < layers->size(); ++i )
+  {
+    if ( layers->at( i )->type() == QgsMapLayerType( 0 ) )
+      layer = qobject_cast<QgsVectorLayer *>( layers->at( i ) );
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.cpp:1364]:
```suggestion
  for ( QgsMapLayer *mapLayer : layers )
  {
    QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( mapLayer );
```

File: src/gui/qgsmapcanvas.cpp [hunk 22]
```diff
@@ -1500,6 +1539,46 @@ void QgsMapCanvas::panToSelected( QgsVectorLayer *layer )
   refresh();
 }
 
+void QgsMapCanvas::panToAllSelected( const QList<QgsMapLayer *> *layers )
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.cpp:1542]:
```suggestion
void QgsMapCanvas::panToSelected( const QList<QgsMapLayer *> &layers )
```

File: src/gui/qgsmapcanvas.cpp [hunk 23]
```diff
@@ -1500,6 +1539,46 @@ void QgsMapCanvas::panToSelected( QgsVectorLayer *layer )
...
+{
+  QgsVectorLayer *layer;
+
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.cpp:1545]:
```suggestion
```

File: src/gui/qgsmapcanvas.cpp [hunk 24]
```diff
@@ -1500,6 +1539,46 @@ void QgsMapCanvas::panToSelected( QgsVectorLayer *layer )
...
+  QgsRectangle rect;
+  rect.setMinimal();
+  QgsRectangle selectionExtent;
+  selectionExtent.setMinimal();
+
+  for ( int i = 0; i < layers->size(); ++i )
+  {
+    if ( layers->at( i )->type() == QgsMapLayerType( 0 ) )
+      layer = qobject_cast<QgsVectorLayer *>( layers->at( i ) );
```
Comment (nyalldawson) [inline src/gui/qgsmapcanvas.cpp:1554]:
```suggestion
  for ( QgsMapLayer *mapLayer : layers )
  {
    QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( mapLayer );
```

File: python/gui/auto_generated/qgsmapcanvas.sip.in [hunk 25]
```diff
@@ -285,6 +286,13 @@ Centers canvas extent to feature ids
     void panToSelected( QgsVectorLayer *layer = 0 );
 %Docstring
 Pan to the selected features of current (vector) layer keeping same extent.
+%End
+
+    void panToAllSelected( const QList<QgsMapLayer *> *layers );
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsmapcanvas.sip.in:291]:
```suggestion
    void panToSelected( const QList<QgsMapLayer *> &layers );
```

File: python/gui/auto_generated/qgsmapcanvas.sip.in [hunk 26]
```diff
@@ -285,6 +286,13 @@ Centers canvas extent to feature ids
...
+%Docstring
+Pan to the combined extent of the selected features of all provided (vector) layers.
+
+:param layers: A list of layers
 %End
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsmapcanvas.sip.in:296]:
```suggestion

.. versionadded:: 3.18
%End
```

File: python/gui/auto_generated/qgsmapcanvas.sip.in [hunk 27]
```diff
@@ -952,6 +960,13 @@ Zoom out with fixed factor
 Zoom to the extent of the selected features of provided (vector) layer.
 
 :param layer: optionally specify different than current layer
+%End
+
+    void zoomToAllSelected( const QList<QgsMapLayer *> *layers );
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsmapcanvas.sip.in:965]:
```suggestion
    void zoomToSelected( const QList<QgsMapLayer *> &layers );
```

File: python/gui/auto_generated/qgsmapcanvas.sip.in [hunk 28]
```diff
@@ -952,6 +960,13 @@ Zoom out with fixed factor
...
+%Docstring
+Zoom to the combined extent of the selected features of all provided (vector) layers.
+
+:param layers: A list of layers
 %End
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsmapcanvas.sip.in:970]:
```suggestion

.. versionadded:: 3.18
%End
```

**Review by TurboGraphxBeige — COMMENTED** [review]
Thanks for the review

**Comment by nyalldawson** [comment]
Superseded by https://github.com/qgis/QGIS/pull/40857

**Pull request closed without merging**